google/cloud/spanner/internal/spanner_stub.cc

```diff
@@ -0,0 +1,257 @@
+    grpc::ClientContext& client_context,
+    spanner_proto::ReadRequest const& request) {
+  spanner_proto::ResultSet response;
+  grpc::Status grpc_status =
+      grpc_stub_->Read(&client_context, request, &response);
+  if (!grpc_status.ok()) {
+    return GRPCStatusToStatus(grpc_status);
+  }
+  return response;
+}
+
+std::unique_ptr<grpc::ClientReaderInterface<spanner_proto::PartialResultSet>>
+DefaultSpannerStub::StreamingRead(grpc::ClientContext& client_context,
+                                  spanner_proto::ReadRequest const& request) {
+  return grpc_stub_->StreamingRead(&client_context, request);
+}
+
+StatusOr<spanner_proto::Transaction> DefaultSpannerStub::BeginTransaction(
+    grpc::ClientContext& client_context,
+    spanner_proto::BeginTransactionRequest const& request) {
+  spanner_proto::Transaction response;
+  grpc::Status grpc_status =
+      grpc_stub_->BeginTransaction(&client_context, request, &response);
+  if (!grpc_status.ok()) {
+    return GRPCStatusToStatus(grpc_status);
+  }
+  return response;
+}
+
+StatusOr<spanner_proto::CommitResponse> DefaultSpannerStub::Commit(
+    grpc::ClientContext& client_context,
+    spanner_proto::CommitRequest const& request) {
+  spanner_proto::CommitResponse response;
+  grpc::Status grpc_status =
+      grpc_stub_->Commit(&client_context, request, &response);
+  if (!grpc_status.ok()) {
+    return GRPCStatusToStatus(grpc_status);
+  }
+  return response;
+}
+
+Status DefaultSpannerStub::Rollback(
+    grpc::ClientContext& client_context,
+    spanner_proto::RollbackRequest const& request) {
+  google::protobuf::Empty response;
+  grpc::Status grpc_status =
+      grpc_stub_->Rollback(&client_context, request, &response);
+  return GRPCStatusToStatus(grpc_status);
+}
+
+StatusOr<spanner_proto::PartitionResponse> DefaultSpannerStub::PartitionQuery(
+    grpc::ClientContext& client_context,
+    spanner_proto::PartitionQueryRequest const& request) {
+  spanner_proto::PartitionResponse response;
+  grpc::Status grpc_status =
+      grpc_stub_->PartitionQuery(&client_context, request, &response);
+  if (!grpc_status.ok()) {
+    return GRPCStatusToStatus(grpc_status);
+  }
+  return response;
+}
+
+StatusOr<spanner_proto::PartitionResponse> DefaultSpannerStub::PartitionRead(
+    grpc::ClientContext& client_context,
+    spanner_proto::PartitionReadRequest const& request) {
+  spanner_proto::PartitionResponse response;
+  grpc::Status grpc_status =
+      grpc_stub_->PartitionRead(&client_context, request, &response);
+  if (!grpc_status.ok()) {
+    return GRPCStatusToStatus(grpc_status);
+  }
+  return response;
+}
+
+}  // namespace
+
+SpannerStub::~SpannerStub() = default;
+
+std::shared_ptr<SpannerStub> CreateDefaultSpannerStub(
+    ClientOptions const& options) {
+  auto channel =
+      grpc::CreateChannel(options.admin_endpoint(), options.credentials());
+  auto spanner_grpc_stub = spanner_proto::Spanner::NewStub(channel);
```

Comment:
Should the NewStub argument be `std::move(channel)`? Even if `CreateChannel` returns a `std::shared_ptr`, moving is more efficient if it's no longer needed in the current scope.

Reply:
`NewStub()` takes a `std::shared_ptr<grpc::Channel> const&`  a `std::move()` here would not help :disappointed:

Reply:
Oh, hmm. Thanks for pointing that out. Well, I guess this API did exist before C++11, and at that point passing shared_ptrs by value everywhere involved lots of unnecessary atomic ops, so passing by const-ref was probably better. In a C++11-world, they may have chosen differently.

**No change necessary here** :-)

[But while we're on the topic, I like that as a caller I kinda don't need to care about how the function accepts an argument. As a caller I only need to know if I am finished with a variable, in which case a `std::move()` is always safe (and never less efficient). And if I'm not finished with a variable, then a `std::move()` is never the right thing. This is nice because I can reason about the code locally without caring about the called function's actual signature. It also allows my code to magically become a bit more efficient if one day the function's signature changes to instead accept a by-value shared_ptr<T>.]